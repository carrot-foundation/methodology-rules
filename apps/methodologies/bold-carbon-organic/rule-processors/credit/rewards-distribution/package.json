{
  "name": "@carrot-fndn/methodologies/bold-carbon-organic/rule-processors/credit/rewards-distribution",
  "version": "0.0.0"
}
