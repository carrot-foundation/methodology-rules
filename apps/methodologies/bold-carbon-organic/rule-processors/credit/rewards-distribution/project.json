{
  "name": "methodologies-bold-carbon-organic-rule-processors-credit-rewards-distribution",
  "$schema": "../../../../../../node_modules/nx/schemas/project-schema.json",
  "tags": ["type:app", "app:any", "stack:any", "methodology:bold"],
  "sourceRoot": "apps/methodologies/bold-carbon-organic/rule-processors/credit/rewards-distribution/src",
  "projectType": "application",
  "targets": {
    "build-lambda": {},
    "lint": {},
    "package-lambda": {},
    "upload-lambda": {},
    "test": {},
    "ts": {}
  }
}
