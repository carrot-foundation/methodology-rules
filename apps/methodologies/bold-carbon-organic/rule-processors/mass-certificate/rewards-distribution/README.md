<div align="center">

# Rewards Distribution

Methodology: **BOLD-CARBON-ORGANIC**

[![GitHub Actions Workflow Status](https://img.shields.io/github/actions/workflow/status/carrot-foundation/methodology-rules/check-and-deploy.yaml)](https://github.com/carrot-foundation/smaug/actions)

</div>

## ðŸ“„ Description

Calculate rewards percentage rights for the participants that contributed to the Circular Economy cycle in this document.

### ðŸ“‚ Implementation

- **[Main Implementation File](./src/lib/rewards-distribution.processor.ts)**

### ðŸ‘¥ Contributors

[![AMarcosCastelo](https://images.weserv.nl/?url=avatars.githubusercontent.com/u/43973049?v=4&h=60&w=60&fit=cover&mask=circle&maxage=7d)](https://github.com/AMarcosCastelo)
[![andtankian](https://images.weserv.nl/?url=avatars.githubusercontent.com/u/12521890?v=4&h=60&w=60&fit=cover&mask=circle&maxage=7d)](https://github.com/andtankian)
[![cris-santos](https://images.weserv.nl/?url=avatars.githubusercontent.com/u/7927374?v=4&h=60&w=60&fit=cover&mask=circle&maxage=7d)](https://github.com/cris-santos)
[![gabrielsl96](https://images.weserv.nl/?url=avatars.githubusercontent.com/u/49005645?v=4&h=60&w=60&fit=cover&mask=circle&maxage=7d)](https://github.com/gabrielsl96)
[![GLGuilherme](https://images.weserv.nl/?url=avatars.githubusercontent.com/u/26340386?v=4&h=60&w=60&fit=cover&mask=circle&maxage=7d)](https://github.com/GLGuilherme)
[![sangalli](https://images.weserv.nl/?url=avatars.githubusercontent.com/u/11515359?v=4&h=60&w=60&fit=cover&mask=circle&maxage=7d)](https://github.com/sangalli)

## ðŸ”‘ License

[License](https://github.com/carrot-foundation/methodology-rules/blob/main/LICENSE)
