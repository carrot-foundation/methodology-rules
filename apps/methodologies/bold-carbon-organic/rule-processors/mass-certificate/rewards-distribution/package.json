{
  "name": "@carrot-fndn/methodologies/bold-carbon-organic/rule-processors/mass-certificate/rewards-distribution",
  "version": "0.0.0"
}
