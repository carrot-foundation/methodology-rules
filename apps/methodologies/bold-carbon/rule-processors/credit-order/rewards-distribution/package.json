{
  "name": "@carrot-fndn/methodologies/bold-carbon/rule-processors/credit-order/rewards-distribution",
  "version": "0.0.0"
}
