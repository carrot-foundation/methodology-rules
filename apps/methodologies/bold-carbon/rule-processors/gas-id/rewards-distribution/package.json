{
  "name": "@carrot-fndn/methodologies/bold-carbon/rule-processors/gas-id/rewards-distribution",
  "version": "0.0.0"
}
