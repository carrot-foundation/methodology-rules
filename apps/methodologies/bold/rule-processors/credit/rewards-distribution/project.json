{
  "name": "methodologies-bold-rule-processors-credit-rewards-distribution",
  "$schema": "../../../../../../node_modules/nx/schemas/project-schema.json",
  "tags": [
    "app:any",
    "methodology:bold",
    "processor:credit",
    "stack:any",
    "type:app"
  ],
  "sourceRoot": "apps/methodologies/bold/rule-processors/credit/rewards-distribution/src",
  "projectType": "app",
  "targets": {
    "build-lambda": {},
    "lint": {},
    "package-lambda": {},
    "upload-lambda": {},
    "test": {},
    "ts": {}
  }
}
