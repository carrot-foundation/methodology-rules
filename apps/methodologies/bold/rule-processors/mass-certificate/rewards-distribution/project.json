{
  "name": "methodologies-bold-rule-processors-certificate-rewards-distribution",
  "$schema": "../../../../../../node_modules/nx/schemas/project-schema.json",
  "tags": [
    "app:any",
    "methodology:bold",
    "processor:certificate",
    "stack:any",
    "type:app"
  ],
  "sourceRoot": "apps/methodologies/bold/rewards-distribution/src",
  "projectType": "app",
  "targets": {
    "build-lambda": {},
    "lint": {},
    "package-lambda": {},
    "upload-lambda": {},
    "test": {},
    "ts": {}
  }
}
