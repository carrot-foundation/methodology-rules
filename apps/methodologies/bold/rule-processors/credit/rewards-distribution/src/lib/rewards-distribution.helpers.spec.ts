import type { NonEmptyArray } from '@carrot-fndn/shared/types';

import {
  type CertificateReward,
  DocumentEventActorType,
  type RewardsDistributionActorType,
} from '@carrot-fndn/methodologies/bold/recycling/organic/types';
import { splitBigNumberIntoParts } from '@carrot-fndn/shared/helpers';
import { stubBigNumber } from '@carrot-fndn/shared/testing';
import { faker } from '@faker-js/faker';
import BigNumber from 'bignumber.js';
import { random, validate } from 'typia';

import type {
  Participant,
  ResultContentWithMassValue,
  RewardsDistribution,
} from './rewards-distribution.types';

import {
  AMOUNT_DECIMALS,
  PERCENTAGE_DECIMALS,
  calculateRewardsDistribution,
  formatPercentage,
  roundAmount,
} from './rewards-distribution.helpers';
import { stubUnitPrice } from './rewards-distribution.stubs';
import { RESULT_CONTENT_WITH_MASS_VALUE_STUB } from './rewards-distribution.test.data';
import { assertExpectedRewardsDistribution } from './rewards-distribution.test.helpers';

describe('Rewards Distribution Helpers', () => {
  describe('formatPercentage', () => {
    it(`should return correct percentage divided by 100 and rounded to ${PERCENTAGE_DECIMALS} decimals`, () => {
      const value = stubBigNumber();
      const totalValue = value.plus(stubBigNumber());

      const percentage = formatPercentage(value, totalValue);

      expect(Number(percentage)).toBeLessThan(100);
      expect(BigNumber(percentage).decimalPlaces()).toBeLessThanOrEqual(
        PERCENTAGE_DECIMALS,
      );
    });

    it('should return 0 if the total value is 0', () => {
      const value = stubBigNumber();
      const totalValue = new BigNumber(0);

      const percentage = formatPercentage(value, totalValue);

      expect(percentage).toBe('0');
    });
  });

  describe('roundAmount', () => {
    it.each([
      ['13.6898828123', '13.689882', 'should round down normally'],
      ['13.6898885999', '13.689888', 'should round down at boundary'],
      ['0.00000123', '0.000001', 'should handle very small amounts'],
      ['999999.9999999', '999999.999999', 'should handle large amounts'],
    ])('case: %s -> %s (%s)', (input, expected) => {
      expect(roundAmount(new BigNumber(input)).toString()).toBe(expected);
      expect(roundAmount(new BigNumber(input)).decimalPlaces()).toBe(
        AMOUNT_DECIMALS,
      );
    });
  });

  describe('calculateRewardsDistribution', () => {
    it('should throw error if two different NETWORK actors are present', () => {
      expect(() =>
        calculateRewardsDistribution(faker.number.float(), [
          {
            massValue: stubBigNumber(),
            resultContent: {
              certificateRewards: [
                random<Participant>(),
                random<Participant>(),
              ].map((participant) => ({
                actorType: DocumentEventActorType.NETWORK,
                participant,
                percentage: '50',
              })) as never,
              massRewards: [] as never,
            },
          },
        ]),
      ).toThrow('Can only have one network actor');
    });

    it('should not throw error if two same NETWORK actors are present', () => {
      const actor = {
        actorType: DocumentEventActorType.NETWORK,
        participant: random<Participant>(),
        percentage: '50',
      };
      const result = calculateRewardsDistribution(faker.number.float(), [
        {
          massValue: stubBigNumber(),
          resultContent: {
            certificateRewards: [actor, actor] as never,
            massRewards: [] as never,
          },
        },
      ]);

      expect(validate<RewardsDistribution>(result)).toPassTypiaValidation();
    });

    it('should match the snapshot', () => {
      const unitPrice = 0.000_001;
      const result = calculateRewardsDistribution(
        unitPrice,
        RESULT_CONTENT_WITH_MASS_VALUE_STUB,
      );

      assertExpectedRewardsDistribution(unitPrice, result);
      expect(result).toMatchSnapshot();
    });

    it('should throw error if NETWORK actors are not present', () => {
      expect(() =>
        calculateRewardsDistribution(faker.number.float(), []),
      ).toThrow('Network actor not found');
    });

    it('should correctly calculate the rewards distribution', () => {
      const unitPrice = stubUnitPrice();
      const actorTypes: NonEmptyArray<RewardsDistributionActorType> = [
        DocumentEventActorType.APPOINTED_NGO,
        DocumentEventActorType.HAULER,
        DocumentEventActorType.INTEGRATOR,
        DocumentEventActorType.METHODOLOGY_AUTHOR,
        DocumentEventActorType.METHODOLOGY_DEVELOPER,
        DocumentEventActorType.NETWORK,
        DocumentEventActorType.PROCESSOR,
        DocumentEventActorType.RECYCLER,
        DocumentEventActorType.SOURCE,
      ];

      const participants: Record<RewardsDistributionActorType, Participant> =
        Object.fromEntries(
          actorTypes.map((actorType) => [actorType, random<Participant>()]),
        ) as never;

      const resultContentsWithMassValue: ResultContentWithMassValue[] = [];
      const certificatesMassValue = [stubBigNumber(), stubBigNumber()];

      for (const massValue of certificatesMassValue) {
        const percentages = splitBigNumberIntoParts(
          new BigNumber(100),
          actorTypes.length,
        );

        const certificateRewards: CertificateReward[] = actorTypes.map(
          (actorType, index) => {
            const percentage = percentages[index]!;

            return {
              actorType,
              participant: participants[actorType],
              percentage: percentage.toString(),
            };
          },
        );

        resultContentsWithMassValue.push({
          massValue,
          resultContent: {
            certificateRewards: certificateRewards as never,
            massRewards: [] as never,
          },
        });
      }

      const result = calculateRewardsDistribution(
        unitPrice,
        resultContentsWithMassValue,
      );

      expect(result.actors.length).toBe(actorTypes.length);
      assertExpectedRewardsDistribution(unitPrice, result);
    });
  });
});
