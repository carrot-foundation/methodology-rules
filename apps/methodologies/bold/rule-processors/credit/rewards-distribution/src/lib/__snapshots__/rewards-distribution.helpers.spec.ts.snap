// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rewards Distribution Helpers calculateRewardsDistribution should match the snapshot 1`] = `
{
  "actors": [
    {
      "actorType": "HAULER",
      "amount": "0.00008",
      "participant": {
        "id": "ighgvmph",
        "name": "okhwrgefhcnxyaaoekkm",
      },
      "percentage": "0.0000007999",
    },
    {
      "actorType": "INTEGRATOR",
      "amount": "0.00009",
      "participant": {
        "id": "ptmeahooevs",
        "name": "i",
      },
      "percentage": "0.0000008999",
    },
    {
      "actorType": "RECYCLER",
      "amount": "0.000008",
      "participant": {
        "id": "ygioqpnmjcak",
        "name": "iasf",
      },
      "percentage": "0.0000000799",
    },
    {
      "actorType": "NETWORK",
      "amount": "0.000034",
      "participant": {
        "id": "ttewegtjrygioqpnmjcak",
        "name": "wupdjxggyfsxb",
      },
      "percentage": "0.0000003399",
    },
    {
      "actorType": "PROCESSOR",
      "amount": "0.00001",
      "participant": {
        "id": "ttewegtjrygioqpnmjcak",
        "name": "wupdjxggyfsxb",
      },
      "percentage": "0.0000000999",
    },
    {
      "actorType": "RECYCLER",
      "amount": "0.000012",
      "participant": {
        "id": "ttewegtjrygioqpnmjcak",
        "name": "wupdjxggyfsxb",
      },
      "percentage": "0.0000001199",
    },
    {
      "actorType": "HAULER",
      "amount": "0.000033",
      "participant": {
        "id": "edge_actor1",
        "name": "Edge Actor 1",
      },
      "percentage": "0.0000003299",
    },
    {
      "actorType": "INTEGRATOR",
      "amount": "0.000033",
      "participant": {
        "id": "edge_actor2",
        "name": "Edge Actor 2",
      },
      "percentage": "0.0000003299",
    },
    {
      "actorType": "RECYCLER",
      "amount": "0.000033",
      "participant": {
        "id": "edge_actor3",
        "name": "Edge Actor 3",
      },
      "percentage": "0.0000003299",
    },
    {
      "actorType": "HAULER",
      "amount": "0.00025",
      "participant": {
        "id": "sum100_actor1",
        "name": "Sum100 Actor 1",
      },
      "percentage": "0.0000024999",
    },
    {
      "actorType": "RECYCLER",
      "amount": "0.00025",
      "participant": {
        "id": "sum100_actor2",
        "name": "Sum100 Actor 2",
      },
      "percentage": "0.0000024999",
    },
    {
      "actorType": "INTEGRATOR",
      "amount": "0",
      "participant": {
        "id": "boundary_actor1",
        "name": "Boundary Actor 1",
      },
      "percentage": "0",
    },
    {
      "actorType": "HAULER",
      "amount": "0",
      "participant": {
        "id": "boundary_actor2",
        "name": "Boundary Actor 2",
      },
      "percentage": "0",
    },
    {
      "actorType": "PROCESSOR",
      "amount": "9999.999999",
      "participant": {
        "id": "boundary_actor3",
        "name": "Boundary Actor 3",
      },
      "percentage": "99.99999167",
    },
  ],
  "massTotalValue": "10000000832.152999",
  "remainder": {
    "amount": "0.000012",
    "percentage": "0.0000001199",
  },
  "unitPrice": 0.000001,
}
`;
